src/backends/interflop-cancellation/interflop_cancellation.c

```diff
@@ -0,0 +1,280 @@
+  (int)bit_a_b(_V_>>_M_,0,_E_-1) - ((1<<_E_-1)-2);})
+
+#define inexact(X,S) (X + pow(2,exp(X)-S) * _mca_rand())
+
+static void _set_mca_seed(int choose_seed, uint64_t seed) {
+  if (choose_seed) {
+    tinymt64_init(&random_state, seed);
+  } else {
+    const int key_length = 3;
+    uint64_t init_key[key_length];
+    struct timeval t1;
+    gettimeofday(&t1, NULL);
+    /* Hopefully the following seed is good enough for Montercarlo */
+    init_key[0] = t1.tv_sec;
+    init_key[1] = t1.tv_usec;
+    init_key[2] = getpid();
+    tinymt64_init_by_array(&random_state, init_key, key_length);
+  }
+}
+
+/****************** CANCELLATION DETECTION FUNCTIONS *************************
+ *  Compute the difference between the max of both operands and the exposant 
+ * of the result to find the size of the cancellation
+ ****************************************************************************/
+
+#define detect(X,Y,Z) ((int)max(exp(X),exp(Y))-exp(Z))
+
+/******************** MCA ARITHMETIC FUNCTIONS ********************
+ * The following set of functions perform the MCA operation. Operands
+ * are first converted to quad  format (GCC), inbound and outbound
+ * perturbations are applied using the _mca_inexact function, and the
+ * result converted to the original format for return
+ *******************************************************************/
+
+// perform_bin_op: applies the binary operator (op) to (a) and (b)
+// and stores the result in (res)
+
+#define mant(X) ((int)(sizeof(X)*8-(6+3*(sizeof(X) >> 2)-1)-1))
+
+#define cancell(X,Y,Z) ({                                         \
+  int cancellation = detect(X,Y,Z);                               \
+  if(cancellation >= MCALIB_T){                                   \
+    if(WARN){                                                     \
+      printf("cancellation of size %d detected\n", cancellation);}\
+      int exp = (mant(Z)+1) - cancellation;                       \
+      Z = inexact(Z,exp);                                         \
+  }                                                               \
+  Z;})
+
+/************************* FPHOOKS FUNCTIONS *************************
+ * These functions correspond to those inserted into the source code
+ * during source to source compilation and are replacement to floating
+ * point operators
+ **********************************************************************/
+
+static void _interflop_add_float(float a, float b, float *c, void *context) {
+  *c = a + b;
+}
+
+static void _interflop_sub_float(float a, float b, float *c, void *context) {
+  *c = a - b;
+  *c = cancell(a, b, *c);
+}
+
+static void _interflop_mul_float(float a, float b, float *c, void *context) {
+  *c = a * b;
+}
+
+static void _interflop_div_float(float a, float b, float *c, void *context) {
+  *c = a / b;
+}
+
+static void _interflop_add_double(double a, double b, double *c,
+                                  void *context) {
+  *c = a + b;
```

Comment:
Same **bug than before**, a and b can be of different sign. In this case cancell should be called.

Reply:
When b is negative and a is positive we agree ,but when a is negative and b positive, it still an addition , isn't it ?

Reply:
In both cases a cancellation can happen.